platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/resolve/service/internal/DefaultInjectedClasspathPluginResolver.java

```diff
@@ -35,29 +41,69 @@
 import java.util.Collection;
 import java.util.stream.Collectors;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 public class DefaultInjectedClasspathPluginResolver implements ClientInjectedClasspathPluginResolver, PluginResolver {
 
     private final ClassPath injectedClasspath;
-    private final PluginRegistry pluginRegistry;
+    private final FileCollectionFactory fileCollectionFactory;
+    private final ScriptClassPathResolver scriptClassPathResolver;
+    private final ClassLoaderScope parentScope;
+    private final PluginInspector pluginInspector;
+    private volatile PluginRegistry pluginRegistry;
 
-    public DefaultInjectedClasspathPluginResolver(ClassLoaderScope parentScope, CachedClasspathTransformer classpathTransformer, PluginInspector pluginInspector, ClassPath injectedClasspath, InjectedClasspathInstrumentationStrategy instrumentationStrategy) {
+    public DefaultInjectedClasspathPluginResolver(
+        ClassLoaderScope parentScope,
+        ScriptClassPathResolver scriptClassPathResolver,
+        FileCollectionFactory fileCollectionFactory,
+        PluginInspector pluginInspector,
+        ClassPath injectedClasspath,
+        InjectedClasspathInstrumentationStrategy instrumentationStrategy
+    ) {
+        this.parentScope = parentScope;
+        this.pluginInspector = pluginInspector;
         this.injectedClasspath = injectedClasspath;
-        ClassPath cachedClassPath = classpathTransformer.transform(injectedClasspath, instrumentationStrategy.getTransform());
-        this.pluginRegistry = new DefaultPluginRegistry(pluginInspector,
-            parentScope.createChild("injected-plugin", null)
-                .local(cachedClassPath)
-                .lock()
-        );
+        this.fileCollectionFactory = fileCollectionFactory;
+        this.scriptClassPathResolver = scriptClassPathResolver;
+        maybeReportAgentUsageWithTestKitProblem(instrumentationStrategy);
+    }
+
+    /**
+     * InstrumentationStrategy will report a problem if the agent is used with TestKit, see ConfigurationCacheInjectedClasspathInstrumentationStrategy class.
+     */
+    private static void maybeReportAgentUsageWithTestKitProblem(InjectedClasspathInstrumentationStrategy instrumentationStrategy) {
+        instrumentationStrategy.getTransform();
     }
 
     @Override
     public void collectResolversInto(Collection<? super PluginResolver> dest) {
         dest.add(this);
     }
 
+    @Override
+    public void prepareClassPath(ScriptHandlerInternal scriptHandlerInternal) {
+        if (pluginRegistry == null) {
```

Comment:
The fact that this method only needs to be called once at _some_ point before we resolve plugins makes me think this detached configuration factory should be a constructor arg instead of passed into this method. In practice it is a dependency on the class. 

Is there a way we can play with service initialization ordering and such to be able to pass this into the constructor?

Reply:
Good point, I will try to make it work.

Reply:
Yeah I can see this one being a little harder. Worth looking into but no worries if it’s not straightforward

Reply:
Ok, I renamed `prepareClassPath` to `preparePluginRegistry` and made it private. Now `preparePluginRegistry` is called when we do `collectResovlersInto`. It's not ideal since now `collectResolversInto` name doesn't tell the whole story, but it is better.

To maybe make that a bit nicer we could add another layer of abstraction and create `DefaultInjectedClasspathPluginResolverFactory` that is injected in to `PluginResolverFactory`. And then could create a new initialized `DefaultInjectedClasspathPluginResolver` in `addDefaultResolvers` via `DefaultInjectedClasspathPluginResolverFactory.createResolver(ScriptHandler)`. Maybe that would be a nicer approach to not abuse `collectResolversInto` for initalization.


I wasn't able to remove `ScriptHandlerInternal` as a parameter, since DependencyHandler is a parameter to:
```
scriptClassPathResolver.prepareClassPath(configuration, dependencies);
```
that we call in the `preparePluginRegistry`. Also injecting any configuration factory doesn't work and it doesn't seem to be straightforward as you assumed.


Any thoughts?

Reply:
Ok, forget about the previous comment. I think I find a way: I now pass `Factory<DependencyResolutionServices>` to `DefaultInjectedClasspathPluginResolver`, that it can be used when creating a plugin registry.